src/cpp/fastdds/domain/DomainParticipantImpl.hpp

```diff
@@ -611,6 +682,7 @@ class DomainParticipantImpl
                 const fastrtps::types::TypeInformation& type_information) override;
 
         DomainParticipantImpl* participant_;
+        int callback_counter_ = 0;
```

Comment:
We should assert `callback_counter_ == 0` on destruction too.

Reply:
I changed the condition to `!(callback_counter_ > 0)` because that's what we really want to assert. The counter should only be `-1` if listener is `nullptr`, or `0` otherwise.

Reply:
Absolutely right!